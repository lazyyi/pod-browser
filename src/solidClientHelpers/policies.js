/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import { getSourceUrl } from "@inrupt/solid-client";
import { sharedStart } from "./utils";
import { getContainerUrl, joinPath } from "../stringHelpers";

const POLICIES_CONTAINER = "pb_policies/";

export function getPoliciesContainerUrl(podRootUri) {
  return joinPath(podRootUri, POLICIES_CONTAINER);
}

export function getPolicyUrl(resource, policiesContainerUrl) {
  const resourceUrl = getSourceUrl(resource);
  const rootUrl = policiesContainerUrl.substr(
    0,
    policiesContainerUrl.length - POLICIES_CONTAINER.length
  );
  const matchingStart = sharedStart(resourceUrl, rootUrl);
  const path = `${resourceUrl.substr(matchingStart.length)}.ttl`;
  return joinPath(getPoliciesContainerUrl(matchingStart), path);
}

export function getResourcePoliciesContainerPath(
  resource,
  policiesContainerUrl
) {
  return getContainerUrl(getPolicyUrl(resource, policiesContainerUrl));
}

export function getPolicyResourceUrl(
  resource,
  policiesContainerUrl,
  policyName
) {
  const resourceUrl = getSourceUrl(resource);
  const rootUrl = policiesContainerUrl.substr(
    0,
    policiesContainerUrl.length - POLICIES_CONTAINER.length
  );
  const matchingStart = sharedStart(resourceUrl, rootUrl);
  const path = `${resourceUrl.substr(
    matchingStart.length
  )}.ttl.${policyName}.ttl`;
  return joinPath(getPoliciesContainerUrl(matchingStart), path);
}
