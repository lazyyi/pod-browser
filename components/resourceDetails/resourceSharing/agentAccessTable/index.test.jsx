/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React from "react";
import userEvent from "@testing-library/user-event";
import { waitFor } from "@testing-library/dom";
import { renderWithTheme } from "../../../../__testUtils/withTheme";
import AgentAccessTable from "./index";
import { createAccessMap } from "../../../../src/solidClientHelpers/permissions";
import usePolicyPermissions from "../../../../src/hooks/usePolicyPermissions";
import usePermissionsWithProfiles from "../../../../src/hooks/usePermissionsWithProfiles";

jest.mock("../../../../src/hooks/usePolicyPermissions");
const mockedUsePolicyPermissions = usePolicyPermissions;

jest.mock("../../../../src/hooks/usePermissionsWithProfiles");
const mockedUsePermissionsWithProfiles = usePermissionsWithProfiles;

const permissions = [
  {
    acl: createAccessMap(true, true, false, false),
    webId: "https://example1.com/profile/card#me",
    profile: {
      avatar: null,
      name: "Example 1",
      types: ["https://schema.org/Person"],
    },
  },
  {
    acl: createAccessMap(true, true, false, false),
    webId: "https://example2.com/profile/card#me",
    profile: {
      avatar: null,
      name: "Example 2",
      types: ["https://schema.org/Person"],
    },
  },
  {
    acl: createAccessMap(true, true, false, false),
    webId: "https://example3.com/profile/card#me",
    profile: {
      avatar: null,
      name: "Example 3",
      types: ["https://schema.org/Person"],
    },
  },
  {
    acl: createAccessMap(true, true, false, false),
    webId: "https://example4.com/profile/card#me",
    profile: {
      avatar: null,
      name: "Example 4",
      types: ["https://schema.org/Person"],
    },
  },
];

describe("AgentAccessTable", () => {
  it("renders an empty list of permissions if there are no permissions", () => {
    const type = "editors";
    mockedUsePolicyPermissions.mockReturnValue({
      data: [],
      mutate: jest.fn(),
    });
    mockedUsePermissionsWithProfiles.mockReturnValue({
      permissionsWithProfiles: [],
    });
    const { asFragment } = renderWithTheme(<AgentAccessTable type={type} />);

    expect(asFragment()).toMatchSnapshot();
  });
  it("renders an empty list of permissions if permissions are unavailable", () => {
    const type = "editors";
    mockedUsePolicyPermissions.mockReturnValue({
      error: "error",
      data: undefined,
      mutate: jest.fn(),
    });
    const { asFragment } = renderWithTheme(<AgentAccessTable type={type} />);

    expect(asFragment()).toMatchSnapshot();
  });
  it("does not render table at all if there aren't any permissions for a custom policy", () => {
    const type = "viewAndAdd";
    mockedUsePolicyPermissions.mockReturnValue({
      data: [],
      mutate: jest.fn(),
    });
    const { asFragment } = renderWithTheme(<AgentAccessTable type={type} />);

    expect(asFragment()).toMatchSnapshot();
  });
  it("renders a list of permissions", async () => {
    mockedUsePolicyPermissions.mockReturnValueOnce({
      data: [
        {
          acl: createAccessMap(true, true, false, false),
          webId: "https://example1.com/profile/card#me",
          profile: {
            avatar: null,
            name: "Example 1",
            types: ["https://schema.org/Person"],
          },
        },
      ],
      mutate: jest.fn(),
    });
    mockedUsePermissionsWithProfiles.mockReturnValue({
      permissionsWithProfiles: [
        {
          acl: createAccessMap(true, true, false, false),
          webId: "https://example1.com/profile/card#me",
          profile: {
            avatar: null,
            name: "Example 1",
            types: ["https://schema.org/Person"],
          },
        },
      ],
    });
    const type = "editors";
    const { asFragment, queryAllByRole } = renderWithTheme(
      <AgentAccessTable type={type} />
    );

    expect(queryAllByRole("cell")).toHaveLength(1);
    expect(asFragment()).toMatchSnapshot();
  });
  it("shows all permissions when clicking 'show all' button", async () => {
    mockedUsePolicyPermissions.mockReturnValueOnce({
      data: permissions,
      mutate: jest.fn(),
    });
    mockedUsePermissionsWithProfiles.mockReturnValue({
      permissionsWithProfiles: permissions,
    });
    const type = "editors";
    const { queryAllByRole, getByTestId } = renderWithTheme(
      <AgentAccessTable type={type} />
    );
    const button = getByTestId("show-all-button");
    expect(queryAllByRole("cell")).toHaveLength(3);
    userEvent.click(button);
    expect(queryAllByRole("cell")).toHaveLength(4);
  });
  it("shows first 3 permissions by default and when clicking the 'hide' button", async () => {
    mockedUsePolicyPermissions.mockReturnValueOnce({
      data: permissions,
      mutate: jest.fn(),
    });
    mockedUsePermissionsWithProfiles.mockReturnValue({
      permissionsWithProfiles: permissions,
    });
    const type = "editors";
    const { queryAllByRole, getByTestId } = renderWithTheme(
      <AgentAccessTable type={type} permissions={permissions} />
    );
    const showAllButton = getByTestId("show-all-button");
    expect(queryAllByRole("cell")).toHaveLength(3);
    userEvent.click(showAllButton);
    expect(queryAllByRole("cell")).toHaveLength(4);
    const hideButton = getByTestId("hide-button");
    userEvent.click(hideButton);
    expect(queryAllByRole("cell")).toHaveLength(3);
  });
  it("renders a search box which filters by name or webId", () => {
    mockedUsePolicyPermissions.mockReturnValueOnce({
      data: permissions,
      mutate: jest.fn(),
    });
    mockedUsePermissionsWithProfiles.mockReturnValue({
      permissionsWithProfiles: permissions,
    });
    const type = "editors";
    const { getByTestId, queryByText } = renderWithTheme(
      <AgentAccessTable type={type} permissions={permissions} />
    );
    const searchInput = getByTestId("search-input");
    userEvent.type(searchInput, "2");
    expect(queryByText("Example 4")).toBeNull();
    expect(queryByText("Example 2")).not.toBeNull();
  });
  // TODO: this will change once we have groups and a new way of filtering when we have groups
  it("renders a set of tabs which filter by Group type", () => {
    const permissionsWithTypes = [
      {
        acl: createAccessMap(true, true, false, false),
        webId: "https://example1.com/profile/card#me",
        profile: {
          avatar: null,
          name: "Example 1",
          types: ["https://schema.org/Person"],
        },
      },
      {
        acl: createAccessMap(true, true, false, false),
        webId: "https://example1.com/profile/card#me",
        profile: {
          avatar: null,
          name: "Not a person",
          types: ["Something else"],
        },
      },
    ];

    mockedUsePolicyPermissions.mockReturnValue({
      data: permissionsWithTypes,
      mutate: jest.fn(),
    });
    mockedUsePermissionsWithProfiles.mockReturnValue({
      permissionsWithProfiles: permissionsWithTypes,
    });

    const type = "editors";
    const { getByTestId, queryByText } = renderWithTheme(
      <AgentAccessTable type={type} />
    );
    const tabPeople = getByTestId("tab-people");
    const tabGroups = getByTestId("tab-groups");
    userEvent.click(tabPeople);
    expect(queryByText("Example 1")).not.toBeNull();
    expect(queryByText("Not a person")).toBeNull();
    userEvent.click(tabGroups);
    expect(queryByText("No groups found")).not.toBeNull();
    expect(queryByText("Example 1")).toBeNull();
  });
  it("renders a set of tabs which filter by People type", () => {
    const permissionsWithTypes = [
      {
        acl: createAccessMap(true, true, false, false),
        webId: "https://example1.com/profile/card#me",
        profile: {
          avatar: null,
          name: "Example 1",
          types: ["Not a Person"],
        },
      },
      {
        acl: createAccessMap(true, true, false, false),
        webId: "https://example1.com/profile/card#me",
        profile: {
          avatar: null,
          name: "Example 2",
          types: ["Not a Person"],
        },
      },
    ];

    mockedUsePolicyPermissions.mockReturnValue({
      data: permissionsWithTypes,
      mutate: jest.fn(),
    });
    mockedUsePermissionsWithProfiles.mockReturnValue({
      permissionsWithProfiles: permissionsWithTypes,
    });

    const type = "editors";
    const { getByTestId, queryByText } = renderWithTheme(
      <AgentAccessTable type={type} />
    );
    const tabPeople = getByTestId("tab-people");
    const tabGroups = getByTestId("tab-groups");
    userEvent.click(tabPeople);
    waitFor(() => {
      expect(queryByText("No people found")).not.toBeNull();
      expect(queryByText("Not a person")).toBeNull();
      userEvent.click(tabGroups);
      expect(queryByText("Example 1")).not.toBeNull();
      expect(queryByText("Example 2")).not.toBeNull();
    });
  });
});
