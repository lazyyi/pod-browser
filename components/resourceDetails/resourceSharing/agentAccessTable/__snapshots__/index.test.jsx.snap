// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AgentAccessTable does not render table at all if there aren't any permissions for a custom policy 1`] = `<DocumentFragment />`;

exports[`AgentAccessTable renders a list of permissions 1`] = `
<DocumentFragment>
  <div
    class="PodBrowser-root PodBrowser-root PodBrowser-accordion PodBrowser-expanded PodBrowser-rounded PodBrowser-rounded PodBrowser-elevation1 PodBrowser-rounded"
    data-testid="agent-access-table"
  >
    <div
      class="PodBrowser-headerContainer"
    >
      <i
        class="PodBrowser-icon-editor PodBrowser-icon PodBrowser-iconEditor"
      />
      <div
        class="PodBrowser-textContainer"
      >
        <div
          class="PodBrowser-titleAndButtonContainer"
        >
          <p
            class="PodBrowser-title"
          >
            Editors
          </p>
          <button
            class="PodBrowser-button PodBrowser-button--text"
            data-testid="add-agent-button"
            iconbefore="add"
          >
            <span
              class="PodBrowser-icon-add PodBrowser-button__icon PodBrowser-button__icon--before"
            />
            <span
              class="PodBrowser-button__text"
            >
              Edit Editors
            </span>
          </button>
        </div>
        <span
          class="PodBrowser-description"
        >
          <p>
            <b>
              Can 
            </b>
            view, edit and delete this resource
          </p>
        </span>
      </div>
    </div>
    <div
      class="PodBrowser-container PodBrowser-entered"
      style="min-height: 0px;"
    >
      <div
        class="PodBrowser-wrapper"
      >
        <div
          class="PodBrowser-wrapperInner"
        >
          <div
            role="region"
          >
            <div
              class="PodBrowser-permissionsContainer"
            >
              <div
                class="PodBrowser-tabsContainer"
              >
                <div
                  class="PodBrowser-root"
                >
                  <div
                    class="PodBrowser-scroller PodBrowser-fixed"
                    style="overflow: hidden;"
                  >
                    <div
                      aria-label="Permissions Filter tabs"
                      class="PodBrowser-flexContainer"
                      role="tablist"
                    >
                      <button
                        aria-selected="true"
                        class="PodBrowser-root PodBrowser-root PodBrowser-tab PodBrowser-textColorInherit PodBrowser-selected PodBrowser-selected"
                        data-testid="tab-all"
                        role="tab"
                        tabindex="0"
                        type="button"
                      >
                        <span
                          class="PodBrowser-wrapper"
                        >
                          All
                        </span>
                        <span
                          class="PodBrowser-root"
                        />
                      </button>
                      <button
                        aria-selected="false"
                        class="PodBrowser-root PodBrowser-root PodBrowser-tab PodBrowser-textColorInherit"
                        data-testid="tab-people"
                        role="tab"
                        tabindex="-1"
                        type="button"
                      >
                        <span
                          class="PodBrowser-wrapper"
                        >
                          People
                        </span>
                        <span
                          class="PodBrowser-root"
                        />
                      </button>
                      <button
                        aria-selected="false"
                        class="PodBrowser-root PodBrowser-root PodBrowser-tab PodBrowser-textColorInherit"
                        data-testid="tab-groups"
                        role="tab"
                        tabindex="-1"
                        type="button"
                      >
                        <span
                          class="PodBrowser-wrapper"
                        >
                          Groups
                        </span>
                        <span
                          class="PodBrowser-root"
                        />
                      </button>
                    </div>
                    <span
                      class="PodBrowser-root PodBrowser-colorSecondary PodBrowser-indicator PodBrowser-indicator"
                      style="left: 0px; width: 0px;"
                    />
                  </div>
                </div>
              </div>
              <div
                class="PodBrowser-searchBoxContainer"
              >
                <button
                  aria-label="search"
                  class="PodBrowser-root PodBrowser-root"
                  tabindex="0"
                  type="submit"
                >
                  <span
                    class="PodBrowser-label"
                  >
                    <i
                      aria-label="search"
                      class="PodBrowser-icon-search PodBrowser-iconSearch"
                    />
                  </span>
                  <span
                    class="PodBrowser-root"
                  />
                </button>
                <div
                  class="PodBrowser-root PodBrowser-searchInput"
                >
                  <input
                    aria-label="Search by name or WebId"
                    class="PodBrowser-input"
                    data-testid="search-input"
                    placeholder="Search by name or WebId"
                    type="text"
                    value=""
                  />
                </div>
              </div>
              <table
                class="PodBrowser-table PodBrowser-agents-table"
                role="table"
              >
                <tbody
                  class="PodBrowser-table__body"
                  role="rowgroup"
                >
                  <tr
                    class="PodBrowser-table__body-row"
                  >
                    <td
                      class="PodBrowser-table__body-cell PodBrowser-agent-cell"
                    >
                      <div
                        class="PodBrowser-nameAndAvatarContainer"
                      >
                        <div
                          class="PodBrowser-root PodBrowser-circle PodBrowser-avatar PodBrowser-colorDefault"
                        >
                          <svg
                            aria-hidden="true"
                            class="PodBrowser-root PodBrowser-fallback"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
                            />
                          </svg>
                        </div>
                        <p
                          class="PodBrowser-root PodBrowser-detailText PodBrowser-body1 PodBrowser-detailText"
                          data-testid="agent-web-id"
                        >
                          Example 1
                        </p>
                        <button
                          class="PodBrowser-button"
                          data-testid="menu-button"
                          type="button"
                        >
                          <i
                            class="PodBrowser-icon-more"
                          />
                        </button>
                      </div>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`AgentAccessTable renders an empty list of permissions if permissions are unavailable 1`] = `
<DocumentFragment>
  <div
    class="PodBrowser-root PodBrowser-root PodBrowser-accordion PodBrowser-expanded PodBrowser-rounded PodBrowser-rounded PodBrowser-elevation1 PodBrowser-rounded"
    data-testid="agent-access-table"
  >
    <div
      class="PodBrowser-headerContainer"
    >
      <i
        class="PodBrowser-icon-editor PodBrowser-icon PodBrowser-iconEditor"
      />
      <div
        class="PodBrowser-textContainer"
      >
        <div
          class="PodBrowser-titleAndButtonContainer"
        >
          <p
            class="PodBrowser-title"
          >
            Editors
          </p>
          <button
            class="PodBrowser-button PodBrowser-button--text"
            data-testid="add-agent-button"
            iconbefore="add"
          >
            <span
              class="PodBrowser-icon-add PodBrowser-button__icon PodBrowser-button__icon--before"
            />
            <span
              class="PodBrowser-button__text"
            >
              Edit Editors
            </span>
          </button>
        </div>
        <span
          class="PodBrowser-description"
        >
          <p>
            <b>
              Can 
            </b>
            view, edit and delete this resource
          </p>
        </span>
      </div>
    </div>
    <div
      class="PodBrowser-container PodBrowser-entered"
      style="min-height: 0px;"
    >
      <div
        class="PodBrowser-wrapper"
      >
        <div
          class="PodBrowser-wrapperInner"
        >
          <div
            role="region"
          >
            <div
              class="PodBrowser-permissionsContainer"
            >
              <span
                class="PodBrowser-emptyStateTextContainer"
              >
                <p>
                  No editors
                </p>
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`AgentAccessTable renders an empty list of permissions if there are no permissions 1`] = `
<DocumentFragment>
  <div
    class="PodBrowser-root PodBrowser-root PodBrowser-accordion PodBrowser-expanded PodBrowser-rounded PodBrowser-rounded PodBrowser-elevation1 PodBrowser-rounded"
    data-testid="agent-access-table"
  >
    <div
      class="PodBrowser-headerContainer"
    >
      <i
        class="PodBrowser-icon-editor PodBrowser-icon PodBrowser-iconEditor"
      />
      <div
        class="PodBrowser-textContainer"
      >
        <div
          class="PodBrowser-titleAndButtonContainer"
        >
          <p
            class="PodBrowser-title"
          >
            Editors
          </p>
          <button
            class="PodBrowser-button PodBrowser-button--text"
            data-testid="add-agent-button"
            iconbefore="add"
          >
            <span
              class="PodBrowser-icon-add PodBrowser-button__icon PodBrowser-button__icon--before"
            />
            <span
              class="PodBrowser-button__text"
            >
              Edit Editors
            </span>
          </button>
        </div>
        <span
          class="PodBrowser-description"
        >
          <p>
            <b>
              Can 
            </b>
            view, edit and delete this resource
          </p>
        </span>
      </div>
    </div>
    <div
      class="PodBrowser-container PodBrowser-entered"
      style="min-height: 0px;"
    >
      <div
        class="PodBrowser-wrapper"
      >
        <div
          class="PodBrowser-wrapperInner"
        >
          <div
            role="region"
          >
            <div
              class="PodBrowser-permissionsContainer"
            >
              <span
                class="PodBrowser-emptyStateTextContainer"
              >
                <p>
                  No editors
                </p>
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
